pyteal/ast/abi/util_test.py

```diff
@@ -772,3 +772,284 @@ def test_sdk_type_specs_from_signature(sig_str, sig_args, sig_rets):
+                        abi.Tuple3[
+                            abi.PaymentTransaction,
+                            abi.Address,
+                            abi.StaticArray[abi.Byte, Literal[16]],
+                        ],
+                        abi.Transaction,
+                    ]
+                ),
+                abi.type_spec_from_annotation(NamedTDecl),
+            ]
+        ),
+    ]
+    + [
+        SafeBidirectional([spec])
+        for spec in abi.ReferenceTypeSpecs + abi.TransactionTypeSpecs
+    ]
+    + [
+        SafeBidirectional([spec_t()])
+        for spec_t in bfs_on_inheritance(abi.UintTypeSpec)
+        if not isabstract(spec_t)
+    ]
+)
+
+
+@pytest.mark.parametrize("tc", SAFE_BIDIRECTIONAL_TEST_CASES)
+def test_type_spec_is_assignable_safe_bidirectional(tc: SafeBidirectional):
+    assert len(tc.xs) > 0
+    for a in tc.xs:
+        for b in tc.xs:
+            assert abi.type_spec_is_assignable_to(a, b)
+
+
+@pytest.mark.parametrize("ts", bfs_on_inheritance(abi.TypeSpec))
+def test_type_spec_is_assignable_safe_bidirectional_full_coverage(ts: type):
+    def exists_in_safe_bidirectional(_ts: type):
+        for safe_bidirectional in SAFE_BIDIRECTIONAL_TEST_CASES:
+            for t in safe_bidirectional.xs:
+                if type(t) == _ts:
+                    return True
+        return False
+
+    if isabstract(ts):
+        assert not exists_in_safe_bidirectional(ts)
+    else:
+        assert exists_in_safe_bidirectional(ts)
+
+
+class SafeAssignment(NamedTuple):
+    a: abi.TypeSpec
+    bs: list[abi.TypeSpec]
+
+
+SAFE_ASSIGNMENT_TEST_CASES: list[SafeAssignment] = [
+    SafeAssignment(
+        abi.StringTypeSpec(),
+        [abi.DynamicBytesTypeSpec(), abi.DynamicArrayTypeSpec(abi.ByteTypeSpec())],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(NamedTDecl),
+        [
+            abi.type_spec_from_annotation(
+                abi.Tuple3[
+                    abi.Uint64,
+                    abi.Tuple3[
+                        abi.Transaction,
+                        abi.Address,
+                        abi.StaticArray[abi.Byte, Literal[16]],
+                    ],
+                    abi.Transaction,
+                ]
+            ),
+        ],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(
+            abi.Tuple3[
+                abi.Uint64,
+                abi.Tuple3[
+                    abi.PaymentTransaction,
+                    abi.Address,
+                    abi.StaticArray[abi.Byte, Literal[16]],
+                ],
+                abi.PaymentTransaction,
+            ]
+        ),
+        [abi.type_spec_from_annotation(NamedTDecl)],
+    ),
+    SafeAssignment(
+        abi.AddressTypeSpec(),
+        [abi.StaticArrayTypeSpec(abi.ByteTypeSpec(), 32), abi.StaticBytesTypeSpec(32)],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(NamedTComp0),
+        [abi.type_spec_from_annotation(NamedTComp1)],
+    ),
```

Comment:
I'm still uncertain about allowing different `NamedTuple` subclasses to be assignable to each other. We cannot know that one is more general than the other.

Consider this example:

```python
class Point(abi.NamedTuple):
    x: abi.Field[abi.Uint64]
    y: abi.Field[abi.Uint64]

class AccountRecord(abi.NamedTuple):
    algoBalance: abi.Field[abi.Uint64]
    assetBalance: abi.Field[abi.Uint64]
```

While these two `NamedTuples` are assignable under the current implementation, they represent very different pieces of information and it would likely not be safe/intended to allow one to be viewed as the other.

The conservative option at this point would be to disallow different `NamedTuple` subclasses from being assignable to each other.

In terms of the example, I propose that `Point` and `AccountRecord` should be assignable to `abi.Tuple[abi.Uint64, abi.Uint64]`, but not to each other.

Thoughts on this?

Reply:
I pushed something in [d12d158](https://github.com/algorand/pyteal/pull/540/commits/d12d158dc04b27884496eb2a8d245d8a552be2e9) as the first step.

My thinking for the thread above:
> Do we want to have assignability between `NamedTuple`?

Only when the `NamedTupleTypeSpec` are *strictly identical*, or we cannot compare the generality.

> Do we still want assignability between `Tuple` and `NamedTuple`?

My preference is still keep the mutual assignability. I see scenarios that one may want to assign some `Tuple` like stuffs into more structured `NamedTuple`.

> Do we want to consider the child classes of `Point` or `AccountRecord`, i.e., child classes of child class of `NamedTuple`?

I think we want to prohibit the patterns. `NamedTuple` instances must be inherited from `NamedTuple`, and only inherited from `NamedTuple`.

Thoughts on this?

Reply:
> I think we want to prohibit the patterns. `NamedTuple` instances must be inherited from `NamedTuple`, and only inherited from `NamedTuple`.

I definitely agree on this point. I don't think we originally considered this case when we implemented `NamedTuple`, but if you were to subclass a subclass of `NamedTuple` today, you would see incorrect behavior.

Consider this example:

```python
class Point2D(abi.NamedTuple):
    """A point in 2D space."""
    x: abi.Field[abi.Uint64]
    y: abi.Field[abi.Uint64]

class Point3D(Point2D):
    """A point in 3D space."""
    z: abi.Field[abi.Uint64]
```

You would expect `Point3D` to inherit the `x` and `y` properties of `Point2D`, plus its own `z` property, but that's not actually what happens. Instead, `Point3D` would only have the `z` property.

This is because the [`get_annotations`](https://github.com/algorand/pyteal/blob/33a4e5e6691ac64f963bdcde036c2a557cf7945d/pyteal/ast/abi/tuple.py#L562) function we use to find all fields on `NamedTuple` classes _only_ looks at the fields declared in that class and ignores all inherited fields.

Due to this incorrect behavior, I'd argue that the fact you can declare and instantiate `Point3D` at all is a bug and should be prevented.

I think the preferred alternative to this example would be embedding instead of inheriting, e.g.:

```python
class Point3D(Point2D):
    """A point in 3D space."""
    point2d: abi.Field[Point2D]
    z: abi.Field[abi.Uint64]
```

Syntactically this is a little less convenient (you have to go through `point2d` to access at `x` and `y`), but it does convey the same idea.